src/com/amazon/corretto/crypto/provider/EvpKey.java

```diff
@@ -0,0 +1,195 @@
+
+    protected byte[] encoded;
+    protected Integer cachedHashCode;
+
+    private static native void releaseKey(long ptr);
+    private static native byte[] encodePublicKey(long ptr);
+    private static native byte[] encodePrivateKey(long ptr);
+    protected static native byte[] getDerEncodedParams(long ptr);
+
+    EvpKey(final InternalKey key, final EvpKeyType type, final boolean isPublicKey) {
+        this.internalKey = key;
+        this.type = type;
+        this.isPublicKey = isPublicKey;
+    }
+
+    boolean isEphemeral() {
+        return ephemeral;
+    }
+
+    void setEphemeral(final boolean ephemeral) {
+        this.ephemeral = ephemeral;
+    }
+
+    void releaseEphemeral() {
+        if (ephemeral) {
+            destroy();
+        }
+    }
+
+    // @CheckReturnValue // Restore once replacement for JSR-305 available
+    <T, X extends Throwable> T use(final MiscInterfaces.ThrowingLongFunction<T, X> function) throws X {
+        return internalKey.use(function);
+    }
+
+    <X extends Throwable> void useVoid(final MiscInterfaces.ThrowingLongConsumer<X> function) throws X {
+        internalKey.useVoid(function);
+    }
+
+    @Override
+    public String getAlgorithm() {
+        return type.jceName;
+    }
+
+    @Override
+    public String getFormat() {
+        return isPublicKey ? "X.509" :  "PKCS#8";
+    }
+
+    @Override
+    public byte[] getEncoded() {
+        initEncoded();
+        return encoded != null ? encoded.clone() : encoded;
+    }
+
+    protected synchronized void initEncoded() {
+        if (encoded == null) {
+            encoded = isPublicKey ? use(EvpKey::encodePublicKey) : use(EvpKey::encodePrivateKey);
+        }
+    }
+
+    protected <X extends Throwable> BigInteger nativeBN(final MiscInterfaces.ThrowingLongFunction<byte[], X> fn)
+            throws X {
+        byte[] raw = use(fn::apply);
+        return new BigInteger(1, raw);
+    }
+
+    protected <T extends AlgorithmParameterSpec> T nativeParams(final Class<T> paramSpec) {
+        byte[] encodedParams = use(EvpKey::getDerEncodedParams);
+        try {
+            AlgorithmParameters params = AlgorithmParameters.getInstance(type.jceName);
+            params.init(encodedParams);
+            return params.getParameterSpec(paramSpec);
+        } catch (final GeneralSecurityException | IOException ex) {
+            throw new RuntimeCryptoException("Unable to deserialize parameters: "
+                + Base64.getEncoder().encodeToString(encoded), ex);
+        }
+    }
+
+    /**
+     * This method will be called by @{link #destroy()} after calling @{code internalKey.release()}.
+     */
+    protected synchronized void destroyJavaState() {
+        // NOP
+    }
+
+    @Override
+    public boolean equals(final Object obj) {
+        // We try to avoid comparing the encoded values
+        // because it may be slow and may pull secret data into the Java heap
+        if (this == obj) {
+            return true;
+        }
```

Comment:
Should we return `false` when ```obj == null```. This is generally expected for `equals` methods.

> For any non-null reference value x, x.equals(null) should return false.
https://docs.oracle.com/javase/8/docs/api/java/lang/Object.html#equals-java.lang.Object-

Reply:
We do return false when `obj == null` because `null instanceof Clazz` is always false, so we return `false` here:

```java
if (!(obj instanceof Key)) { // Implicit null check
    return false;
}
```

Reply:
You're right.  I was thinking that an expression like `null instanceof Key` would throw a NullPointerException, but that's not the case.  The language specification says that it will always return `false`: https://docs.oracle.com/javase/specs/jls/se11/html/jls-15.html#jls-15.20.2